### PR TITLE
TIMIT recipe

### PR DESCRIPTION
**Pull request by sknadig**
TIMIT recipe with minimal changes to the original scripts. Data preparation derived from kaldi-asr with changes to timit_data_prep.sh to handle both phoneme and char level data preparation for TIMIT.

I had to change text2token.py and data2json.py. We could have a local copy for TIMIT if they're conflicting for other recipes.

### CODE REVIEW

File: egs/timit/asr1/.gitignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+exp
+data
+dump
+fbank
+tensorboard
```
Comment (kan-bayashi) [inline egs/timit/asr1/.gitignore]: These are already included in .gitignore @ root directory. 
Please remove it.

File: egs/timit/asr1/cmd.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+# with slurm.  Different queues are configured differently, with different
+# queue names and different ways of specifying things like memory;
+# to account for these differences you can create and edit the file
+# conf/queue.conf to match your queue's configuration.  Search for
+# conf/queue.conf in http://kaldi-asr.org/doc/queue.html for more information,
+# or search for the string 'default_config' in utils/queue.pl or utils/slurm.pl.
+
+export train_cmd="run.pl --mem 2G"
+export cuda_cmd="run.pl --mem 2G --gpu 1"
+export decode_cmd="run.pl --mem 4G"
```
Comment (kan-bayashi) [inline egs/timit/asr1/cmd.sh]: Please copy cmd.sh from other recipe of espnet.

File: egs/timit/asr1/local/parse_options.sh [hunk 3]
```diff
@@ -0,0 +1,97 @@
+#!/bin/bash
+
```
Comment (kan-bayashi) [inline egs/timit/asr1/local/parse_options.sh]: what is the difference between this script and utils/parse_options.sh?

File: egs/timit/asr1/run.sh [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+do_delta=false
+
+# network architecture
+# encoder related
+etype=blstmp     # encoder architecture type
+elayers=4
+eunits=320
+eprojs=320
+subsample=1_2_2_1_1 # skip every n frame from input to nth layers
+# subsample=1_1_1_1_1
```
Comment (kan-bayashi) [inline egs/timit/asr1/run.sh]: remove unnecessary comment out.

File: egs/timit/asr1/run.sh [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    echo "stage 1: Feature Generation"
+    fbankdir=fbank
+    # Generate the fbank features; by default 80-dimensional fbanks with pitch on each frame
+    for x in test train dev; do
+        steps/make_fbank_pitch.sh --cmd "$train_cmd" --nj 8 --write_utt2num_frames true \
+        data/${x} exp/make_fbank/${x} ${fbankdir}
+    done
+    
+    # make a dev set
+
```
Comment (kan-bayashi) [inline egs/timit/asr1/run.sh]: remove blank line.

File: utils/text2token.py [hunk 6]
```diff
@@ -40,6 +40,8 @@ def main():
                         help='list of non-linguistic symobles, e.g., <NOISE> etc.')
     parser.add_argument('text', type=str, default=False, nargs='?',
                         help='input text')
+    parser.add_argument('--trans_type', '-t', type=str, default="char",
```
Comment (kan-bayashi) [inline utils/text2token.py]: How about add choices=[“char”, “phn”]?

File: utils/text2token.py [hunk 7]
```diff
@@ -86,13 +88,18 @@ def main():
...
+        if("phn" in args.trans_type):
+            a = a.split(" ")
+        else:
+            a = [a[j:j + n] for j in range(0, len(a), n)]
 
         a_flat = []
         for z in a:
             a_flat.append("".join(z))
-
+
```
Comment (kan-bayashi) [inline utils/text2token.py]: remove unnecessary spaces.

File: utils/text2token.py [hunk 8]
```diff
@@ -86,13 +88,18 @@ def main():
                     i += 1
             a = chars
 
-        a = [a[j:j + n] for j in range(0, len(a), n)]
+        if("phn" in args.trans_type):
```
Comment (kan-bayashi) [inline utils/text2token.py]: I think following representation us better:

if args.trans_type == “phn”:

File: utils/text2token.py [hunk 9]
```diff
@@ -86,13 +88,18 @@ def main():
...
+        else:
+            a = [a[j:j + n] for j in range(0, len(a), n)]
 
         a_flat = []
         for z in a:
             a_flat.append("".join(z))
-
+            
         a_chars = [z.replace(' ', args.space) for z in a_flat]
+        if("phn" in args.trans_type):
```
Comment (kan-bayashi) [inline utils/text2token.py]: Same comment.

File: utils/text2token.py [hunk 10]
```diff
@@ -40,6 +40,8 @@ def main():
                         help='list of non-linguistic symobles, e.g., <NOISE> etc.')
     parser.add_argument('text', type=str, default=False, nargs='?',
                         help='input text')
+    parser.add_argument('--trans_type', '-t', type=str, default="char",
+                        help='Transcript type. char/phn')
```
Comment (kan-bayashi) [inline utils/text2token.py]: Could you add short example here in both case?

**Comment by sknadig** [comment]
@kan-bayashi Thank you for the suggestions. Updated accordingly.

File: espnet/bin/asr_train.py [hunk 11]
```diff
@@ -69,7 +69,7 @@ def main(cmd_args):
                         help='Subsample input frames x_y_z means subsample every x frame at 1st layer, '
                              'every y frame at 2nd layer etc.')
     # loss
-    parser.add_argument('--ctc_type', default='warpctc', type=str,
+    parser.add_argument('--ctc_type', default='builtin', type=str,
```
Comment (sw005320) [inline espnet/bin/asr_train.py]: better to keep to use `warpctc` as default.

File: egs/timit/asr1/run.sh [hunk 12]
```diff
@@ -0,0 +1,236 @@
+#!/bin/bash
+
+# Copyright 2017 Johns Hopkins University (Shinji Watanabe)
```
Comment (sw005320) [inline egs/timit/asr1/run.sh]: Change it to you!

**Comment by sw005320** [comment]
Many thanks!

**Pull request merged**